src/new_menu_helpers.c

```diff
@@ -0,0 +1,765 @@
+    }
+}
+
+void ResetTempTileDataBuffers(void)
+{
+    int i;
+    for (i = 0; i < (s32)ARRAY_COUNT(gUnknown_203AB60); i++)
+    {
+        gUnknown_203AB60[i] = NULL;
+    }
+    gUnknown_203AB5C = 0;
+}
+
+bool8 FreeTempTileDataBuffersIfPossible(void)
+{
+    int i;
+
+    if (!IsDma3ManagerBusyWithBgCopy())
+    {
+        if (gUnknown_203AB5C)
+        {
+            for (i = 0; i < gUnknown_203AB5C; i++)
+            {
+                FREE_AND_SET_NULL(gUnknown_203AB60[i]);
+            }
+            gUnknown_203AB5C = 0;
+        }
+        return FALSE;
+    }
+    else
+    {
+        return TRUE;
+    }
+}
+
+void *DecompressAndCopyTileDataToVram(u8 bgId, const void *src, u32 size, u16 offset, u8 mode)
+{
+    u32 sizeOut;
+    if (gUnknown_203AB5C < ARRAY_COUNT(gUnknown_203AB60))
+    {
+        void *ptr = MallocAndDecompress(src, &sizeOut);
+        if (!size)
+            size = sizeOut;
+        if (ptr)
+        {
+            CopyDecompressedTileDataToVram(bgId, ptr, size, offset, mode);
+            gUnknown_203AB60[gUnknown_203AB5C++] = ptr;
+        }
+        return ptr;
+    }
+    return NULL;
+}
+
+void *DecompressAndCopyTileDataToVram2(u8 bgId, const void *src, u32 size, u16 offset, u8 mode)
+{
+    u32 sizeOut;
+    if (gUnknown_203AB5C < ARRAY_COUNT(gUnknown_203AB60))
+    {
+        void *ptr = MallocAndDecompress(src, &sizeOut);
+        if (sizeOut > size)
+            sizeOut = size;
+        if (ptr)
+        {
+            CopyDecompressedTileDataToVram(bgId, ptr, sizeOut, offset, mode);
+            gUnknown_203AB60[gUnknown_203AB5C++] = ptr;
+        }
+        return ptr;
+    }
+    return NULL;
+}
+
+void DecompressAndLoadBgGfxUsingHeap(u8 bgId, const void *src, u32 size, u16 offset, u8 mode)
+{
+    u32 sizeOut;
+    void *ptr = MallocAndDecompress(src, &sizeOut);
+    if (!size)
+        size = sizeOut;
+    if (ptr)
+    {
+        u8 taskId = CreateTask(TaskFreeBufAfterCopyingTileDataToVram, 0);
+        gTasks[taskId].data[0] = CopyDecompressedTileDataToVram(bgId, ptr, size, offset, mode);
+        SetWordTaskArg(taskId, 1, (u32)ptr);
+    }
+}
+
+void DecompressAndLoadBgGfxUsingHeap2(u8 bgId, const void *src, u32 size, u16 offset, u8 mode)
+{
+    u32 sizeOut;
+    void *ptr = MallocAndDecompress(src, &sizeOut);
+    if (sizeOut > size)
+        sizeOut = size;
+    if (ptr)
+    {
+        u8 taskId = CreateTask(TaskFreeBufAfterCopyingTileDataToVram, 0);
+        gTasks[taskId].data[0] = CopyDecompressedTileDataToVram(bgId, ptr, sizeOut, offset, mode);
+        SetWordTaskArg(taskId, 1, (u32)ptr);
+    }
+}
+
+static void TaskFreeBufAfterCopyingTileDataToVram(u8 taskId)
+{
+    if (!CheckForSpaceForDma3Request(gTasks[taskId].data[0]))
+    {
+        Free((void *)GetWordTaskArg(taskId, 1));
+        DestroyTask(taskId);
+    }
+}
+
+void *MallocAndDecompress(const void *src, u32 *size)
+{
+    void *ptr;
+    u8 *sizeAsBytes = (u8 *)size;
+    u8 *srcAsBytes = (u8 *)src;
```

Comment:
No need to typecast a void pointer when assigning it to a non-void pointer variable

Reply:
The problem lying here is that I'm assigning a const ptr to a non-const one, should I modify function signature?

Reply:
No, just declare `srcAsBytes` as `const u8 *`